src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { cloneDeep } from 'lodash';
+
+export enum PERMISSION_TYPE {
+  READ = 'read',
+  WRITE = 'write',
+  MANAGEMENT = 'management',
+  LIBRARY_READ = 'library_read',
+  LIBRARY_WRITE = 'library_write',
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Partial<Record<PERMISSION_TYPE, Principals>>;
+
+const PERMISSION_TYPE_MAP = new Map<string, PERMISSION_TYPE>([
+  ['read', PERMISSION_TYPE.READ],
+  ['write', PERMISSION_TYPE.WRITE],
+  ['management', PERMISSION_TYPE.MANAGEMENT],
+  ['library_read', PERMISSION_TYPE.LIBRARY_READ],
+  ['library_write', PERMISSION_TYPE.LIBRARY_WRITE],
+]);
+
+const addPermissionToPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users && !(principals.users?.length === 1 && principals.users[0] === '*')) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups && !(principals.groups?.length === 1 && principals.groups[0] === '*')) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deletePermissionFromPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const isParamsValid = (
+  permissionType: string,
+  user?: string | string[],
+  group?: string | string[]
+) => {
+  if ((!user && !group) || !permissionType || !PERMISSION_TYPE_MAP.get(permissionType)) {
+    return false;
+  }
+  return true;
+};
+
+export class WorkspacePermissionCheck {
```

Comment:
I am thinking if we can make the ACL class more easy to use.
```
const aclInstance = new ACL(initialACL);
const finalACLJSON = aclInstance
    .add(permissionType, { users: ["user a"] })
    .remove(permissionType, { users: ["user b"] })
    .getJSON();
```

Reply:
One more thing is that if we can rename the class to like `ACL` or anything else, the class has nothing to do with workspace, it is a pure ACL class for other services to operation `ACL` data without knowing the implementation detail.